pkg/kf/commands/group/group.go

```diff
@@ -0,0 +1,160 @@
+	t := template.New("")
+	if templateFuncs == nil {
+		t.Funcs(defaultTemplateFuncs)
+	} else {
+		t.Funcs(*templateFuncs)
+	}
+
+	template.Must(t.Parse(text))
+	return t.Execute(out, data)
+}
+
+// rpad adds padding to the right of a string.
+func rpad(s string, padding int) string {
+	template := fmt.Sprintf("%%-%ds", padding)
+	return fmt.Sprintf(template, s)
+}
+
+/// trimRightSpace trims tailing whitespace.
+func trimRightSpace(s string) string {
+	return strings.TrimRightFunc(s, unicode.IsSpace)
+}
+
+// CalculateMinWidth gets the minimum width required for all command names.
+func CalculateMinWidth(groups CommandGroups) int {
+	minWidth := 0
+	for _, group := range groups {
+		for _, c := range group.Commands {
+			if c == nil {
+				continue
+			}
+			if len(c.Name()) > minWidth {
+				minWidth = len(c.Name())
+			}
+		}
+	}
+	return minWidth
+}
+
+// PrintTrimmedMultilineString does just that.
+func PrintTrimmedMultilineString(str string, out io.Writer) {
+	if str == "" {
+		return
+	}
+
+	for _, line := range strings.Split(str, "\n") {
+		fmt.Fprintln(out, strings.TrimSpace(line))
+	}
+}
+
+// CommandGroupUsageFunc returns a UsageFunc a root level command can use.
+func CommandGroupUsageFunc(groups CommandGroups, templateFuncs *template.FuncMap) func(*cobra.Command) error {
+	return func(command *cobra.Command) error {
+		out := command.OutOrStdout()
+		fmt.Fprintln(out)
+		return PrintTemplate(out, command.UsageTemplate(), command, templateFuncs)
+	}
+}
+
+// CommandGroupHelpFunc returns a HelpFunc a root level command can use.
+func CommandGroupHelpFunc(rootCommand *cobra.Command, groups CommandGroups, templateFuncs *template.FuncMap) func(*cobra.Command, []string) {
+	return func(command *cobra.Command, args []string) {
+		out := command.OutOrStdout()
+
+		// not the root level, use the default template
+		if rootCommand != command {
+			err := PrintTemplate(out, command.HelpTemplate(), command, templateFuncs)
+			if err != nil {
+				panic(fmt.Sprintf("Error printing help: %v", err))
+			}
+			return
+		}
+
+		PrintTrimmedMultilineString(command.Long, out)
+		fmt.Fprintln(out)
+
+		minWidth := CalculateMinWidth(groups)
+		// add 2 for the prefix spaces, 1 for the padding between cols
+		minWidth += 3
+
+		tabout := tabwriter.NewWriter(out, minWidth, 8, 1, ' ', 0)
+		defer tabout.Flush()
+		for _, group := range groups {
+			fmt.Fprintln(tabout, group.Name)
+			for _, c := range group.Commands {
+				fmt.Fprintf(tabout, "  %s\t%s\n", c.Name(), c.Short)
+			}
+			fmt.Fprintln(tabout)
+		}
+		fmt.Fprintln(tabout, "Usage:")
+		fmt.Fprintf(tabout, "  %s [flags] COMMAND\n\n", command.CommandPath())
+		fmt.Fprintf(tabout, "Use \"%s COMMAND --help\" for more information about a given command.\n", command.CommandPath())
+	}
+}
+
+// AddCommandGroups adds CommandGroups to a given cobra Command.
+// A side effact of calling this function is that subcommands will no longer be
+// able to use their own usage or help funcs.
```

Comment:
It the second part of this comment still accurate?

Reply:
Yes, because of UsageFunc() https://github.com/spf13/cobra/blob/master/command.go#L340
performing a parent lookup before accessing the struct private usageFunc. As far as I can tell there's no way to unset a parent from a Command.